File: src/main/java/fds/java/events/EventReader.java
package fds.java.events;

import io.nats.streaming.StreamingConnection;
import io.nats.streaming.StreamingConnectionFactory;
import io.nats.streaming.SubscriptionOptions;

import java.io.IOException;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.CountDownLatch;
import java.util.stream.Collectors;
import java.util.stream.Stream;


/**
 * Stub for the Event Reader of the Nats EventStore
 */
public class EventReader {

    final static Set<String> targets= new HashSet<>(
                                Arrays.asList(  "AdultSki",
                                                "WinterActivities",
                                                "DanielBaudSkiGuide")
                                );

    final static void register(final String logLine, final Map<String,Integer> document) {
        targets.forEach( ( target ) -> {
            if ( logLine.contains((target)) )
                document.compute(target, (k, v) -> (v == null) ? 1 : v + 1);
        });
    }


    public static void main(String[] args) throws Exception {

        final Map<String,Integer> document = new HashMap<>();

        // Connect to EventStore
        final String clusterID = "test-cluster";
        final String clientID = "event-reader";
        final StreamingConnectionFactory cf = new StreamingConnectionFactory(clusterID, clientID);
        final StreamingConnection sc = cf.createConnection();

        // Subscribe to the store for events
        final String subject = "BBC7";

        // you may want to remove count down latch for the stream.
        final CountDownLatch doneLatch = new CountDownLatch(1);
        final SubscriptionOptions opts = new SubscriptionOptions.Builder().deliverAllAvailable().build();
        sc.subscribe(subject, evt -> {
            System.out.println(evt);
            doneLatch.countDown();
        } , opts);

        doneLatch.await();
        sc.close();
	}


	// Format a map into a report
    final static String formatDoc(final Map<String,Integer> document ) {
        List<String> rep = new ArrayList<>();
        rep.add("--------------------------------------------------------------------------------");
        rep.add("| Giant Media - Usage Report                                                   |");
        rep.add("--------------------------------------------------------------------------------");
        document.forEach( (k,v) -> rep.add(String.format("| %-49s|%26d |",k,v)));
        rep.add("--------------------------------------------------------------------------------");
        final Stream<Integer> views = document.values().stream();
        rep.add(String.format("| %-49s|%26d |","Total Views", views.reduce(0, Integer::sum)));
        rep.add("--------------------------------------------------------------------------------");
        return rep.stream().collect(Collectors.joining("\n"));
    }


    // Write a String into a named file
    final static void writeToFile(final String filename, final String content) throws IOException {
        Files.write(Paths.get(filename), content.getBytes(StandardCharsets.UTF_8));
    }

    // Read a String from a named file
    final static String readFromFile(final String filename) throws IOException {
        return new String(Files.readAllBytes(Paths.get(filename)),StandardCharsets.UTF_8);
    }


}

File: src/main/scala/fds/scala/events/EventReader.scala
package fds.scala.events


import java.nio.charset.StandardCharsets
import java.nio.file.{Files, Paths}
import java.util._
import java.util.concurrent.CountDownLatch

import io.nats.streaming.{Message, StreamingConnection, StreamingConnectionFactory, SubscriptionOptions}

import scala.collection.mutable


object EventReader {

  private[events] val targets  =
    "AdultSki" :: "WinterActivities" :: "DanielBaudSkiGuide" :: Nil

  private[events] def register(logLine: String, document: Map[String, Int]): Unit = {
    targets.foreach((target: String) => {
      if (logLine.contains(target))
        document.compute(target, (k, v) => if (v.equals(null)) 1 else v + 1)
    })
  }

  @throws[Exception]
  def main(args: Array[String]): Unit = {

    val document = new HashMap[String, Int]

    // Connect to EventStore
    val clusterID: String = "test-cluster"
    val clientID: String = "event-reader"
    val cf: StreamingConnectionFactory = new StreamingConnectionFactory(clusterID, clientID)
    val sc: StreamingConnection = cf.createConnection

    // Subscribe to the store for events
    val subject: String = "BBC7"
    // you may want to remove count down latch for the stream.
    val doneLatch: CountDownLatch = new CountDownLatch(1)

    //  Many subscribe options @see https://nats.io/documentation/writing_applications/subscribing/
    val opts: SubscriptionOptions = new SubscriptionOptions.Builder().deliverAllAvailable.build

    sc.subscribe(subject, (evt: Message) => {
      System.out.println(evt)
      doneLatch.countDown
    }, opts)

    // wait for a message
    doneLatch.await
    // tidy up the connection
    sc.close
  }

  import scala.collection.JavaConverters._


  // Format the map into a report
  private[events] def formatDoc(document: Map[String, Int]): String = {
    val rep  = mutable.ArrayBuffer[String]()
    rep += "--------------------------------------------------------------------------------"
    rep += "| Giant Media - Usage Report                                                   |"
    rep += "--------------------------------------------------------------------------------"
    rep ++= document.asScala.map{ case (k,v) => String.format("| %-49s|%26d |", k, new Integer(v)) }
    rep += "--------------------------------------------------------------------------------"
    val views = document.asScala.values.sum
    rep += String.format("| %-49s|%26d |", "Total Views", new Integer(views))
    rep += "--------------------------------------------------------------------------------"
    return rep.mkString("\n")
  }

  // Write a String into a named file
  private[events] def writeToFile(filename: String, content: String): Unit = {
    Files.write(Paths.get(filename), content.getBytes(StandardCharsets.UTF_8))
  }

  // Read a String from a named file
  private[events] def readFromFile(filename: String) = {
    new  String(Files.readAllBytes(Paths.get(filename)), StandardCharsets.UTF_8)
  }


}